instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java

```diff
@@ -0,0 +1,243 @@
+  }
+
+  /**
+   * Configure the {@link JarAnalyzer}. If not called, {@link #handle(ProtectionDomain)} is a noop.
+   *
+   * @throws IllegalStateException if called multiple times
+   */
+  public synchronized void configure(int jarsPerSecond) {
+    if (this.jarsPerSecond != null) {
+      throw new IllegalStateException("JarAnalyzer has already been configured");
+    }
+    this.jarsPerSecond = jarsPerSecond;
+    this.handler = this::handleInternal;
+  }
+
+  /**
+   * Install {@link OpenTelemetry} and start processing archives if {@link #configure(int)} was
+   * called.
+   *
+   * @throws IllegalStateException if called multiple times
+   */
+  public synchronized void maybeInstall(OpenTelemetry unused) {
+    if (worker != null) {
+      throw new IllegalStateException("JarAnalyzer has already been installed");
+    }
+    if (this.jarsPerSecond == null) {
+      return;
+    }
+    // TODO(jack-berg): Use OpenTelemetry to obtain EventEmitter when event API is stable
+    EventEmitter eventEmitter =
+        GlobalEventEmitterProvider.get()
+            .eventEmitterBuilder(JmxRuntimeMetricsUtil.getInstrumentationName())
+            .setInstrumentationVersion(JmxRuntimeMetricsUtil.getInstrumentationVersion())
+            .setEventDomain(EVENT_DOMAIN_PACKAGE)
+            .build();
+    this.worker = new Worker(eventEmitter, toProcess, jarsPerSecond);
+    Thread workerThread =
+        new DaemonThreadFactory(JarAnalyzer.class.getSimpleName() + "_WorkerThread")
+            .newThread(worker);
+    workerThread.start();
+  }
+
+  /**
+   * Identify the archive (JAR or WAR) associated with the {@code protectionDomain} and queue it to
+   * be processed if its the first time we've seen it.
+   *
+   * <p>NOTE: does nothing if {@link #configure(int)} has not been called.
+   */
+  void handle(ProtectionDomain protectionDomain) {
+    this.handler.accept(protectionDomain);
+  }
+
+  private void handleInternal(ProtectionDomain protectionDomain) {
+    if (protectionDomain == null) {
+      return;
+    }
+    CodeSource codeSource = protectionDomain.getCodeSource();
+    if (codeSource == null) {
+      return;
+    }
+    URL archiveUrl = codeSource.getLocation();
+    if (archiveUrl == null) {
+      return;
+    }
+    URI locationUri;
+    try {
+      locationUri = archiveUrl.toURI();
+    } catch (URISyntaxException e) {
+      logger.log(Level.WARNING, "Unable to get URI for code location URL: " + archiveUrl, e);
+      return;
+    }
+
+    if (!seenUris.add(locationUri)) {
+      return;
+    }
+    if ("jrt".equals(archiveUrl.getProtocol())) {
+      logger.log(Level.FINEST, "Skipping processing for java runtime module: " + archiveUrl);
+      return;
+    }
+    String file = archiveUrl.getFile();
+    if (file.endsWith("/")) {
+      logger.log(Level.FINEST, "Skipping processing non-archive code location: " + archiveUrl);
+      return;
+    }
+    if (!file.endsWith(JAR_EXTENSION) && !file.endsWith(WAR_EXTENSION)) {
```

Comment:
What about `ear`? Some people might still use that

Reply:
Oh didn't know about ear. Let me see if I can't get the test module to build an `.ear` so I can confirm the logic works.

Reply:
The ear isn't usually an issue because app servers will extract them during the deploy. The exception is jboss that uses a custom url protocol that can represent nested archives. Similarly war files are almost always extracted to support `ServletContext.getRealPath`, I think only weblogic may deploy (there is a checkbox in the weblogic admin console) from packaged war, but that won't affect class loading as they extract WEB-INF/lib and also package classes from WEB-INF/classes into a jar.